File: src/main/java/com/kigya/patterns/java/observer/publisher/WeatherData.java
package com.kigya.patterns.java.observer.publisher;

import com.kigya.patterns.java.observer.subscribers.Observer;
import lombok.Getter;

import java.util.ArrayList;
import java.util.List;

@Getter
public class WeatherData implements Subject {
    private final List<Observer> observers;
    private float temperature;
    private float humidity;
    private float pressure;

    public WeatherData() {
        observers = new ArrayList<>();
    }

    public void registerObserver(Observer o) {
        observers.add(o);
    }

    public void removeObserver(Observer o) {
        observers.remove(o);
    }

    public void notifyObservers() {
        observers.forEach(observer -> observer.update(temperature, humidity, pressure));
    }

    public void measurementsChanged() {
        notifyObservers();
    }

    public void setMeasurements(float temperature, float humidity, float pressure) {
        this.temperature = temperature;
        this.humidity = humidity;
        this.pressure = pressure;
        measurementsChanged();
    }

}

File: src/main/java/com/kigya/patterns/kotlin/observer/publisher/WeatherData.kt
package com.kigya.patterns.kotlin.observer.publisher

import com.kigya.patterns.kotlin.observer.subscibers.Observer

class WeatherData : Subject<WeatherData> {

    var temperature: Float = 0f
        private set

    var humidity: Float = 0f
        private set

    var pressure: Float = 0f
        private set

    private val observers: ArrayList<Observer<WeatherData>> = arrayListOf()

    fun setMeasurements(temperature: Float, humidity: Float, pressure: Float) {
        this.temperature = temperature
        this.humidity = humidity
        this.pressure = pressure
        onMeasurementsChanged()
    }

    private fun onMeasurementsChanged() = notifyObservers()

    override fun registerObserver(observer: Observer<WeatherData>) {
        observers.add(observer)
    }

    override fun removeObserver(observer: Observer<WeatherData>) {
        observers.remove(observer)
    }

    override fun notifyObservers() = observers.forEach {
        it.update(this)
    }
}